[Merged by Bors] - feat: Add Partitioner trait with SiphashRoundRobinPartitioner impl

nicholastmosher opened the pull request:
Closes #838 

- Introduces `Partitioner` trait to enable new implementations in the future
- Adds `SiphashRoundRobinPartitioner` as a first implementor of `Partitioner`
  - Uses Siphash-2-4 on records that have keys
  - Uses round-robin to assign partitions to records with no keys
- Adds a `StoreContext<TopicSpec>` to `MetdataStores` for syncing topic metadata
- Implements watch requests on SPU for Topic and SPG metadata

sehz commented:
Can you link this to issue?

simlay commented on examples/01-produce-key-value/src/main.rs:
```diff
@@ -23,17 +23,22 @@
 
 async fn produce_key_value() -> Result<(), fluvio::FluvioError> {
-    let producer = fluvio::producer("key-value").await?;
+    let producer = fluvio::producer("twop").await?;
```
`twop`?

nicholastmosher replied:
Whoops! I changed that for testing out two partitions. I'll switch it back

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +180,131 @@ impl TopicProducer {
+///
+/// See [`SiphashRoundRobinPartitioner`] for a reference implementation.
+trait Partitioner {
+    fn partition(&self, keys: &[Option<&[u8]>], partition_count: i32) -> Vec<i32>;
```
Is partitioner stateless or stateful?

nicholastmosher replied:
The partitioner should have the opportunity to be stateful. In the case of `SiphashRoundRobinPartitioner`, it keeps track of the partition index which should be used next in the round-robin selection.

sehz replied:
Why is partition count need to be passed every time?

nicholastmosher replied:
That is a good point, we could add the `partition_count` as a piece of state in the implementor of `Partitioner`. I think one relevant question is: Is it possible for the partition_count to change, and if so, how often is that likely to ever happen? If it is very likely to happen frequently, it might make sense to keep it as an argument. If it is unlikely to happen, we may want to add an auxiliary function to the trait, something like `fn resize_partitions(&self, partition_count: i32)`. I think using an argument is just simpler though, but let me know what you think.

sehz replied:
We should make it general.  Kafka partitioning uses combination of key, values and metadata.  We need metadata in case we support rack assignment

nicholastmosher replied:
I see. In that case, I think what we should do is make the trait start out like this:

```rust
trait Partitioner {
    fn partition(&self, keys: &[Option<&[u8]>]) -> i32; // Note removal of partition_count
}
```

Then, what we can do is leave it up to the individual Partitioner implementations to require the particular type of configuration they need in order to make decisions about how to go about partitioning. So for our initial partitioner, we would have:

```rust
struct SiphashRoundRobinPartitioner {
    partition_count: i32,
    // ...
}

impl SiphashRoundRobinPartitioner  {
    pub fn new(partition_count: i32) -> Self { ... }
}
```

And later on we could update that constructor to take some more general type, e.g. something like `ClusterConfig` that may have more info about the topics and partitions.

Additionally, in the future we could add a new function to the `Partitioner` trait that allows us to send configuration updates to the implementors:

```rust
trait Partitioner<T=()> {
    fn partition(...) { ... }
    fn update_config(&mut self, config: T);
}
```

sehz replied:
Why does this need to take array of keys? Partitioning only needed for per record. So it could be like 
```
trait Partitioner {
    fn partition(&self, key: Option<Bytes>,value: Bytes) -> i32; 
    fn update_config(&mut self, config: T
}
```

nicholastmosher replied:
Right. The reason is because I had originally designed the Partitioner to use interior mutability, i.e. if it needed to keep track of something like a round-robin index, it would keep it in an `Arc<Mutex<i32>>` or something. I did not want to need to lock and unlock a mutex for every key, so I made the API accept a batch of keys at once.

However, now that we're doing mutability at a higher level, I can change this back to one key at a time 👍

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +180,131 @@ impl TopicProducer {
+        let some_u64_mod_i32 = some_u64 % (i32::MAX as u64);
+        assert!(some_u64_mod_i32 < i32::MAX as u64);
+        let some_i32 = some_u64_mod_i32 as i32;
+        println!(
```
prefer not to have println in test code.  if need to diagnose internal state, put as debug or should do another assert

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +180,131 @@ impl TopicProducer {
+
+    /// Ensure that feeding keyless records one-at-a-time does not assign the same partition
+    #[test]
+    fn test_round_robin_individual() {
```
can you change this test to match what is intended to be? like `test_round_robin_unique_partition`

nicholastmosher replied:
Sure thing, I'll clean this up and add some assertions that are clearer

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +180,131 @@ impl TopicProducer {
+        for _ in 0..100 {
+            let p = partitioner.partition(&[None], partition_count)[0];
+            println!("Partition: {}", p);
+            assert_ne!(p, prev_partition);
```
not clear behavior of this test. it might be easier if you test individual state

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +180,131 @@ impl TopicProducer {
+        let mut prev_partition = -1;
+        for _ in 0..10 {
+            let ps = partitioner.partition(&batch, partition_count);
+
```
again, test behavior are hard to understand

sehz commented on src/client/src/sync/controller.rs:
```diff
@@ -97,7 +95,7 @@ where
             }
 
-            select! {
+            tokio::select! {
```
not fan of adding prefix

sehz commented on src/client/src/sync/controller.rs:
```diff
@@ -111,21 +109,20 @@ where
                                 Ok(update) => {
                                     if let Err(err) = self.process_updates(update).await {
-                                        error!("processing updates: {}", err);
+                                        error!("Error processing updates: {}", err);
```
this is already error, so marking as error seems to be redundant.  However, welcome making error clear...

sehz commented on src/client/src/sync/controller.rs:
```diff
@@ -111,21 +109,20 @@ where
                         },
                         Some(Err(err)) => {
-                            error!("{} error receiving, end, {}", S::LABEL, err);
+                            error!("Error receiving response, ending: {}", err);
```
ditto

sehz commented on src/client/src/sync/controller.rs:
```diff
@@ -137,55 +134,33 @@ where
-
-                    }
-                }).collect();
+        debug!(count = updates.all.len(), "Syncing updates:");
```
can you comment on what changes happen here?

nicholastmosher replied:
Sure. Previously, there was a really big if-statement like

```rust
if !updates.all.is_empty() {
    // ...
}
```

which was causing a lot of rightward drift. So all I did was remove that if-statement and shift everything inside back to the left, and at the top I added the opposite if-statement to short-circuit the function:

```rust
if updates.all.is_empty() {
    // ...
    return Ok(());
}
```

So it just made the code easier to read while I was trying to figure out what was going wrong.

sehz commented on src/client/src/sync/mod.rs:
```diff
@@ -1,164 +1,159 @@
+use crate::metadata::spu::SpuSpec;
+use crate::metadata::core::MetadataItem;
+
+pub(crate) type CacheMetadataStoreObject<S> = MetadataStoreObject<S, AlwaysNewContext>;
```
Can you minimize change here so it's to easier to see diff? If more make it cleaner, can you do that in separate PR?

nicholastmosher replied:
I remember removing an inline module, I'll go back and undo that to reduce noise in the diff.

nicholastmosher commented on src/client/src/producer.rs:
```diff
@@ -48,7 +59,32 @@ impl TopicProducer {
+        skip(self, value),
+        fields(topic = %self.topic),
+    )]
+    pub async fn send_keyless<V>(&self, value: V) -> Result<(), FluvioError>
```
This is meant for the use case of: I want to send one record with no key. Previously, we would use `send_record`, but the problem with that is that it asked for an explicit partition number, which we no longer want to do. This is essentially the same function as `send_record` but minus the `partition` argument.

sehz replied:
let's not add more API unless it is absolutely necessary.

nicholastmosher replied:
Ok, so if we do not include this `send_keyless` API, then the only way to send a single record with no key is to do:

```rust
producer.send_all(Some((None, "Hello!"))).await?;
```

which I think is much more difficult to write and certainly not at all intuitive if you are a user searching for that functionality. If we do include `send_keyless`, users can use it to do the same thing:

```rust
producer.send_keyless("Hello!").await?;
```

So in this case I would actually argue that a function like `send_keyless` is actually very necessary. I am open to suggestions if you have any other preferences for the name of it though.

sehz replied:
Hmm. we still have `send` right? so it can be:
```
producer.send(None,"Hello!").await?;
```

nicholastmosher replied:
You cannot pass `None` to send. This is the API for `send`:

```rust
    pub async fn send<K, V>(&self, key: K, value: V) -> Result<(), FluvioError>
    where
        K: Into<Vec<u8>>,
        V: Into<Vec<u8>>,
    { ... }
```

So you must pass both a key and a value as arguments to send, you cannot use None. I have also tried other designs for this API but there does not seem to be a good solution for an optional key. This is the reason we need a separate API like `send_keyless`.

sehz replied:
Let's move creating new or updating existing API as another issue which doesn't belong here.

For handling none,  I suggest we use custom wrapper struct rather than using `Vec<u8>` directory.  This way we have full control of conversion.

nicholastmosher commented on src/client/src/producer.rs:
```diff
@@ -62,21 +98,69 @@ impl TopicProducer {
+                .push(DefaultBatch::new(records));
+            topic_request.name = self.topic.clone();
+            topic_request.partitions.push(partition_request);
+            request.acks = 1;
```
This is how the producer request was already being built (see the removed diff in `send_records_raw`. If we want to change this we should do it in a new issue

sehz replied:
Got it.  Let's create issue to allow developers to configure producer and profile. We probably need producer builder or something like that.

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +180,128 @@ impl TopicProducer {
+/// - Records with keys get their keys hashed with siphash
+/// - Records without keys get assigned to partitions using round-robin
+struct SiphashRoundRobinPartitioner {
+    index: Arc<RwLock<i32>>,
```
why this need to be in the lock?  A producer doesn't need to be thread safe nor shared

nicholastmosher replied:
> A producer doesn't need to be thread safe nor shared

Well, right now all of our producer's methods take self by immutable reference `&self`:

- `fn send<K, V>(&self, ...)`
- `fn send_keyless<V>(&self, ...)`
- `fn send_all<K, V>(&self, ...)`

And `TopicProducer` actually does implement Send and Sync. So in order to have mutable state we need to put the interior mutability somewhere. I am actually reworking some of this based on your earlier comment, let me ping you back on this when I have tweaked it a bit.

nicholastmosher commented:
Ok, newest changes:

- Updated `Partitioner` trait to look like the following:

```rust
trait Partitioner {
    fn partition(&mut self, keys: &[Option<&[u8]>]) -> Vec<i32>;
    fn update_config(&mut self, config: PartitionerConfig);
}
```

- Added `PartitionerConfig` to hold information such as `partition_count` in a way that lets us expand in the future
- Moved interior mutability out of `Partitioner` itself and up into the `TopicProducer` level

sehz commented on src/client/src/producer.rs:
```diff
@@ -62,21 +100,76 @@ impl TopicProducer {
+
+        for (partition, request) in requests {
+            let replica = ReplicaKey::new(&self.topic, partition);
+            let mut spu_client = self.pool.create_serial_socket(&replica).await?;
```
This seems inefficient, we are creating new client event thought they are same.  Should cache them per SPU given partition

nicholastmosher replied:
Yes, I was thinking about that as well. In order to make that happen, I'll essentially need a map of `SpuId -> [PartitionId]`, so that I could create one request per SPU rather than per partition. I will look today to see how to put that together.

nicholastmosher replied:
Actually, I want a mapping of `PartitionId -> LeaderId`.

I am looking through the `TopicSpec` metadata to try to figure out how to get this mapping, and I have a few questions. 

- I see we can either have an `Assigned` or `Computed` topic, and it seems like the `PartitionMaps` in the  Assigned topic might give me the information that I need
  - I see the `PartitionMaps` is essentially a `[PartitionMap { id: PartitionId, replicas: [ SpuId ] } ]`
  - Does the index into `PartitionMaps` represent the leader ID? It doesn't seem that this would make sense, because then there could only be one partition assigned to a leader.
  - Does the index into `PartitionMaps` actually represent the Partition ID, and the thing called `id: PartitionId` actually represent the leader's SPU ID? I think that would make more sense because then multiple partitions could be assigned to the same leader
- How do we handle Computed topics?

@sehz let me know if I'm on the right track or if I'm way off

sehz replied:
Topics PartitionMap represents allocations. It tells which replica (unit of storage) is assigned to SPU. It doesn't tell you about leader.  That is tracked in the `Partition` spec.

sehz replied:
so topic list tells you where replicas are:
```
flvd topic list
 NAME   TYPE      PARTITIONS  REPLICAS  IGNORE-RACK  STATUS                   REASON 
 test   computed      1          1                   resolution::provisioned   
 test2  computed      1          2                   resolution::provisioned   
```

but partition tells you which are leader:
```
flvd partition list
 TOPIC  PARTITION  LEADER  REPLICAS  RESOLUTION  HW    LEO   LSR  FOLLOWER OFFSETS 
 test2  0          5001    [5002]    Online      1     1     0    [ReplicaStatus { spu: 5002, hw: 0, leo: 0 }] 
 test   0          5001    []        Online      1000  1000  0    [] 
```

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +200,103 @@ impl TopicProducer {
+///
+/// See [`SiphashRoundRobinPartitioner`] for a reference implementation.
+trait Partitioner {
+    fn partition(&mut self, key: Option<&[u8]>) -> i32;
```
See previous comment, this should take on value as well

nicholastmosher replied:
- Which previous comment are you referring to?
- Why should the partitioner accept the value for the record? What is an example use-case that you are imagining?

sehz replied:
It was in previous comment:  https://kafka.apache.org/24/javadoc/?org/apache/kafka/clients/producer/Partitioner.html.  Kafka produce takes on value which make sense.   There is no need to limit to key only.

nicholastmosher replied:
Oh interesting, it looks like they have partitioning strategies that just ignore the key and look at the value contents to find a defined partition https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/UniformStickyPartitioner.java#L25-L36

sehz commented on src/client/src/producer.rs:
```diff
@@ -62,21 +101,86 @@ impl TopicProducer {
+                .or_insert_with(HashMap::new)
+                .entry(partition)
+                .or_insert_with(Vec::new)
+                .push(record);
```
it is important to add unit test for this

nicholastmosher replied:
Added! I factored out the `group_by_spu` function and wrote a pretty good test for it, let me know what you think

sehz reviewed (changes requested):
Good progress.  See other comments regarding moving keyless api as separate issue

sehz commented:
Also, test with replication factor = 2

sehz commented:
Can you add this to changelog?

sehz commented on examples/00-produce/src/main.rs:
```diff
@@ -32,9 +32,9 @@ async fn main() {
-    producer.send_record(&record, 0).await?;
-    println!("{}", &record);
+    let value = "Hello, Fluvio!";
+    producer.send_keyless(value).await?;
```
`keyless` api is still in this PR.  Please move them to separate PR

sehz commented on examples/03-echo/src/main.rs:
```diff
@@ -138,13 +138,13 @@ async fn main() {
             .await?;
     }
-    producer.send_record("Done!", 0).await?;
+    producer.send_keyless("Done!").await?;
```
ditto

sehz commented on src/cli/src/consumer/produce/mod.rs:
```diff
@@ -102,7 +102,7 @@ impl ProduceOpt {
             self.produce_key_value(producer, string).await?;
         } else {
-            producer.send_record(string, 0).await?;
+            producer.send_keyless(string).await?;
```
ditto

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +195,219 @@ impl TopicProducer {
-        let spu_client = self.pool.create_serial_socket(&replica).await?;
+async fn group_by_spu(
+    topic: &str,
+    partitions: &StoreContext<PartitionSpec>,
```
doesn't need dependencies to `StoreContext`, it just need `LocalStore` which assume cache are populated

nicholastmosher replied:
The StoreContext is the type with the `lookup_by_key` method which I'm using to implement this. It is not obvious to me how to go about this with LocalStore, and what is written here is working well

sehz replied:
Actually, this may be good.  If live SPU is not found, it can wait.  Ok. let's leave as is

sehz commented on src/client/src/producer.rs:
```diff
@@ -62,21 +103,79 @@ impl TopicProducer {
+                    .records
+                    .batches
+                    .push(DefaultBatch::new(records));
+                topic_request.partitions.push(partition_request);
```
unit test?

nicholastmosher replied:
Factored out `assemble_requests` method and wrote a test for it

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +195,219 @@ impl TopicProducer {
+        );
+
+        Ok(())
     }
```
need test with replication

nicholastmosher replied:
I think replication tests are more fitting as an integration test or e2e test. Let's make another issue for it

sehz commented on tests/runner/src/tests/smoke/produce.rs:
```diff
@@ -131,7 +131,7 @@ pub async fn produce_message_with_api(
             let len = message.len();
             info!("trying send: {}, iteration: {}", topic_name, i);
-            producer.send_record(message, 0).await.unwrap_or_else(|_| {
+            producer.send_keyless(message).await.unwrap_or_else(|_| {
```
ditto

sehz commented on tests/runner/src/tests/producer_stress.rs:
```diff
@@ -82,7 +82,7 @@ pub async fn run(client: Arc<Fluvio>, mut test_case: TestCase) -> TestResult {
             let message = format!("producer-{} line-{} {}", i, n, long_str.clone());
 
-            p.send_record(message.clone(), 0)
+            p.send_keyless(message)
```
ditto

sehz commented on tests/runner/src/tests/concurrent/producer.rs:
```diff
@@ -18,6 +18,6 @@ pub async fn producer(fluvio: Arc<Fluvio>, option: ConcurrentTestCase, digests:
         let record_digest = hash_record(&record);
         digests.send(record_digest).unwrap();
-        producer.send_record(&record, PARTITION).await.unwrap();
+        producer.send_keyless(record).await.unwrap();
```
ditto

sehz reviewed (changes requested):
Making good progress.   Couple of minor changes needed

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +148,354 @@ impl TopicProducer {
-        debug!("connect to replica leader at: {}", spu_client);
+fn assemble_requests(
+    topic: &str,
+    partitions_by_spu: HashMap<i32, HashMap<i32, Vec<DefaultRecord>>>,
```
Change `i32` to `SpuId`

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +148,354 @@ impl TopicProducer {
+///
+/// See [`SiphashRoundRobinPartitioner`] for a reference implementation.
+trait Partitioner {
+    fn partition(&mut self, key: Option<&[u8]>, value: &[u8]) -> i32;
```
change i32 to SpuId

nicholastmosher replied:
I don't think this is a `SpuId`, this is the Partition. Is there a `PartitionId` type somewhere?

nicholastmosher replied:
There is `fluvio_types::PartitionId` which is `i32`, I can use that

sehz commented on src/client/src/producer.rs:
```diff
@@ -96,82 +148,354 @@ impl TopicProducer {
+/// - Records with keys get their keys hashed with siphash
+/// - Records without keys get assigned to partitions using round-robin
+struct SiphashRoundRobinPartitioner {
+    index: i32,
```
change to `SpuId`

nicholastmosher replied:
Changed to PartitionId ;)

nicholastmosher commented on src/client/src/producer.rs:
```diff
@@ -62,21 +78,57 @@ impl TopicProducer {
+            let mut partitioner = self.partitioner.lock().await;
+            partitioner.update_config(partition_config);
+
+            for record in entries {
```
I tried but I actually cannot. The body of this loop uses the `partitioner`, so if that body is moved into a closure, it then borrows the `partitioner` for longer than allowed (because iterators are lazy and using `.map` simply saves the closure object)

sehz replied:
k. fair enough

sehz reviewed (approved):
LGTM.  Nice work!

sehz commented:
bors r+
